### DISCUSSION TITLE
Customizable Error Handling for findOrFail in Lucid Models

### DISCUSSION THREAD

**Discussion by Natchii59** [adonisjs/core#4692] [Ideas]
### Context
Currently, in AdonisJS, when using the `findOrFail` method on a Lucid model, if the record is not found, a lucid **`E_ROW_NOT_FOUND`** error is thrown by default. While this exception provides useful information, it would be beneficial to allow developers to customize this default error at the model level.

### Proposal
I propose adding a feature that allows setting a default exception for record retrieval methods (like `findOrFail`) directly on the respective model. This configuration would enable specifying a custom exception for cases where the record is not found.

### Example
Here’s an example of what the implementation could look like:

```ts
export default class User extends BaseModel {
  @column({ isPrimary: true })
  declare id: number

  @column()
  declare fullName: string | null

  @column.dateTime({ autoCreate: true })
  declare createdAt: DateTime

  @column.dateTime({ autoCreate: true, autoUpdate: true })
  declare updatedAt: DateTime | null

  // Adding a custom exception for findOrFail
  static failException = UserNotFoundException
}
```

### Benefits
1. Enhanced Customization: Allows developers to better control the execution flow by adapting errors to their specific needs.
2. Readability: Makes the code clearer and more explicit about the types of errors to expect when a record is missing.
3. Reusability: Facilitates consistent error handling and avoids duplicating logic across different parts of the application.

### Conclusion
This feature would provide more flexibility and clarity in error handling within AdonisJS applications. I would be happy to collaborate on the implementation of this proposal or provide any other necessary contribution.

**Comment by Xstoudi** [comment]
I don't think hyper-specializing the exception at model level is useful, especially if the goal is to be able to end up with `<ModelName>NotFoundException` for each model.

However, if you consider it's related to your proposal, I think it would be interesting to be able to recover the model who threw the exception like it's possible in Laravel. Something like:
````ts
if (error.code === 'E_ROW_NOT_FOUND') {
  console.log(error.model.name) // error.model is User model class
}
```
It was already very briefly discussed here : https://github.com/orgs/adonisjs/discussions/3946
